File: tmdb-core/src/main/java/dev/reuss/tmdb/core/http/JavaNetTmdbHttpClient.java
package dev.reuss.tmdb.core.http;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import dev.reuss.tmdb.core.config.TmdbClientConfig;
import dev.reuss.tmdb.core.exception.*;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

/**
 * {@link TmdbHttpClient} implementation based on Java's built-in
 * {@link HttpClient}.
 *
 * <p>This client applies the configured TMDB authentication, builds absolute
 * request URIs from {@link TmdbRequest} instances, sends HTTP GET requests and
 * maps successful JSON responses to the requested Java type.</p>
 *
 * <p>Non-successful HTTP responses are translated into SDK-specific runtime
 * exceptions such as {@link TmdbUnauthorizedException},
 * {@link TmdbNotFoundException}, {@link TmdbRateLimitException} or
 * {@link TmdbServerException}.</p>
 */
public final class JavaNetTmdbHttpClient implements TmdbHttpClient {

    private final TmdbClientConfig config;
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;

    /**
     * Creates a new HTTP client using the given TMDB client configuration.
     *
     * @param config the TMDB client configuration
     * @throws NullPointerException if {@code config} is {@code null}
     */
    public JavaNetTmdbHttpClient(TmdbClientConfig config) {
        this.config = Objects.requireNonNull(config, "TMDB client config must not be null");
        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(config.connectTimeout())
                .build();
        this.objectMapper = new ObjectMapper()
                .findAndRegisterModules()
                .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
    }

    /**
     * Sends a GET request to TMDB and maps the JSON response body to the given type.
     *
     * @param request      the TMDB request to send
     * @param responseType the expected response type
     * @param <T>          the expected response type
     * @return the mapped response body
     * @throws NullPointerException      if {@code request} or {@code responseType} is {@code null}
     * @throws TmdbUnauthorizedException if TMDB returns HTTP 401
     * @throws TmdbNotFoundException     if TMDB returns HTTP 404
     * @throws TmdbRateLimitException    if TMDB returns HTTP 429
     * @throws TmdbServerException       if TMDB returns an HTTP 5xx response
     * @throws TmdbMappingException      if the response body cannot be mapped
     * @throws TmdbException             for network errors, interruptions or other non-successful responses
     */
    @Override
    public <T> T get(TmdbRequest request, Class<T> responseType) {
        Objects.requireNonNull(request, "TMDB request must not be null");
        Objects.requireNonNull(responseType, "Response type must not be null");

        URI uri = buildUri(request);

        HttpRequest httpRequest = HttpRequest.newBuilder(uri)
                .timeout(config.requestTimeout())
                .header("Authorization", config.auth().authorizationHeaderValue())
                .header("Accept", "application/json")
                .GET()
                .build();

        try {
            HttpResponse<String> response = httpClient.send(
                    httpRequest,
                    HttpResponse.BodyHandlers.ofString()
            );

            return handleResponse(response, responseType);

        } catch (IOException e) {
            throw new TmdbClientException("Failed to execute TMDB request", e);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new TmdbClientException("TMDB request was interrupted", e);
        }
    }

    private URI buildUri(TmdbRequest request) {
        String baseUrl = config.baseUrl().endsWith("/")
                ? config.baseUrl().substring(0, config.baseUrl().length() - 1)
                : config.baseUrl();

        String url = baseUrl + request.path();

        if (!request.queryParams().isEmpty()) {
            url += "?" + toQueryString(request.queryParams());
        }

        return URI.create(url);
    }

    private String toQueryString(Map<String, String> queryParams) {
        return queryParams.entrySet()
                .stream()
                .map(entry -> encode(entry.getKey()) + "=" + encode(entry.getValue()))
                .collect(Collectors.joining("&"));
    }

    private String encode(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }

    private <T> T handleResponse(HttpResponse<String> response, Class<T> responseType) {
        int httpStatus = response.statusCode();
        String body = response.body();

        if (httpStatus >= 200 && httpStatus < 300) {
            return mapBody(body, responseType, httpStatus);
        }

        throw mapErrorResponse(httpStatus, body);
    }

    private TmdbApiException mapErrorResponse(int httpStatus, String body) {
        try {
            TmdbErrorResponse errorResponse = objectMapper.readValue(
                    body,
                    TmdbErrorResponse.class
            );

            return TmdbExceptions.from(
                    httpStatus,
                    errorResponse.statusCode(),
                    errorResponse.statusMessage(),
                    body
            );
        } catch (Exception exception) {
            return TmdbExceptions.fromHttpStatus(
                    httpStatus,
                    "TMDB request failed with status code " + httpStatus,
                    body
            );
        }
    }

    private <T> T mapBody(String body, Class<T> responseType, int statusCode) {
        try {
            return objectMapper.readValue(body, responseType);
        } catch (Exception exception) {
            throw new TmdbMappingException(
                    "Failed to map TMDB response to " + responseType.getSimpleName(),
                    exception
            );
        }
    }
}

File: tmdb-core/src/main/kotlin/dev/reuss/tmdb/core/http/JavaNetTmdbHttpClient.kt
package dev.reuss.tmdb.core.http

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.KotlinFeature
import com.fasterxml.jackson.module.kotlin.KotlinModule
import dev.reuss.tmdb.core.config.TmdbClientConfig
import dev.reuss.tmdb.core.exception.TmdbApiException
import dev.reuss.tmdb.core.exception.TmdbClientException
import dev.reuss.tmdb.core.exception.TmdbErrorResponse
import dev.reuss.tmdb.core.exception.TmdbExceptions
import dev.reuss.tmdb.core.exception.TmdbMappingException
import dev.reuss.tmdb.core.metrics.TmdbMetricsRecorder
import org.slf4j.LoggerFactory
import java.io.IOException
import java.net.URI
import java.net.URLEncoder
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.nio.charset.StandardCharsets
import java.time.Duration

/**
 * Default [TmdbHttpClient] based on Java's built-in [HttpClient].
 *
 * The client applies bearer authentication, the configured timeouts, and the
 * default `language` and optional `region`. Request-specific query parameters
 * replace defaults with the same name. Query names and values are UTF-8 encoded.
 *
 * Unknown JSON properties are ignored and JSON `null` collections and maps are
 * mapped to empty Kotlin collections. Non-successful responses, transport
 * failures, and mapping failures are translated to SDK exceptions.
 */
class JavaNetTmdbHttpClient(
    private val config: TmdbClientConfig,
) : TmdbHttpClient {
    private val httpClient: HttpClient =
        HttpClient
            .newBuilder()
            .connectTimeout(config.connectTimeout)
            .build()

    private val objectMapper: ObjectMapper =
        ObjectMapper()
            .registerModule(
                KotlinModule
                    .Builder()
                    .enable(KotlinFeature.NullToEmptyCollection)
                    .enable(KotlinFeature.NullToEmptyMap)
                    .build(),
            ).findAndRegisterModules()
            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)

    private val metricsRecorder: TmdbMetricsRecorder = config.metricsRecorder

    /**
     * Executes a blocking GET request and maps its response body.
     *
     * @throws TmdbApiException when TMDB returns a non-2xx response
     * @throws TmdbClientException when transport fails or the thread is interrupted
     * @throws TmdbMappingException when a 2xx response cannot be mapped
     */
    override fun <T> get(
        request: TmdbRequest,
        responseType: Class<T>,
    ): T {
        log.debug(
            "Sending TMDB request: method=GET, path={}, responseType={}",
            request.path,
            responseType.simpleName,
        )

        val uri = buildUri(request)
        log.trace("Resolved TMDB request URI: {}", uri)

        val httpRequest =
            HttpRequest
                .newBuilder(uri)
                .timeout(config.requestTimeout)
                .header("Authorization", config.auth.authorizationHeaderValue())
                .header("Accept", "application/json")
                .GET()
                .build()

        val method = httpRequest.method()
        val path = request.path
        val startedAt = System.nanoTime()

        metricsRecorder.recordRequestStarted(method, path)

        try {
            val response =
                httpClient.send(
                    httpRequest,
                    HttpResponse.BodyHandlers.ofString(),
                )

            val durationMillis = elapsedMillis(startedAt)

            metricsRecorder.recordRequestFinished(
                method,
                path,
                response.statusCode(),
                Duration.ofMillis(durationMillis),
                responseBytes(response.body()),
            )

            return handleResponse(
                response,
                responseType,
                method,
                path,
                durationMillis,
            )
        } catch (exception: IOException) {
            val duration = Duration.ofNanos(System.nanoTime() - startedAt)

            metricsRecorder.recordRequestFailed(
                method,
                path,
                exception,
                duration,
            )

            throw TmdbClientException(
                "Failed to execute TMDB request",
                exception,
            )
        } catch (exception: InterruptedException) {
            val duration = Duration.ofNanos(System.nanoTime() - startedAt)

            metricsRecorder.recordRequestFailed(
                method,
                path,
                exception,
                duration,
            )

            Thread.currentThread().interrupt()

            throw TmdbClientException(
                "TMDB request was interrupted",
                exception,
            )
        }
    }

    private fun buildUri(request: TmdbRequest): URI {
        val baseUrl = config.baseUrl.removeSuffix("/")
        val queryParams = effectiveQueryParams(request)

        val url =
            buildString {
                append(baseUrl)
                append(request.path)

                if (queryParams.isNotEmpty()) {
                    append("?")
                    append(toQueryString(queryParams))
                }
            }

        return URI.create(url)
    }

    private fun effectiveQueryParams(request: TmdbRequest): Map<String, String> =
        QueryParams
            .create()
            .add("language", config.defaultLanguage.value)
            .add("region", config.defaultRegion?.value)
            .apply {
                request.queryParams.forEach { (name, value) ->
                    add(name, value)
                }
            }.toMap()

    private fun toQueryString(queryParams: Map<String, String>): String =
        queryParams.entries.joinToString("&") { (name, value) ->
            "${encode(name)}=${encode(value)}"
        }

    private fun encode(value: String): String = URLEncoder.encode(value, StandardCharsets.UTF_8)

    private fun <T> handleResponse(
        response: HttpResponse<String>,
        responseType: Class<T>,
        method: String,
        path: String,
        durationMillis: Long,
    ): T {
        val httpStatus = response.statusCode()
        val body = response.body()

        log.debug(
            "Received TMDB response: status={}, path={}, responseType={}, duration={}ms",
            httpStatus,
            path,
            responseType.simpleName,
            durationMillis,
        )

        if (httpStatus in 200..299) {
            return mapBody(
                body,
                responseType,
                method,
                path,
            )
        }

        logNonSuccessfulResponse(
            httpStatus,
            path,
            durationMillis,
        )

        throw mapErrorResponse(httpStatus, body)
    }

    private fun mapErrorResponse(
        httpStatus: Int,
        body: String,
    ): TmdbApiException =
        try {
            val errorResponse =
                objectMapper.readValue(
                    body,
                    TmdbErrorResponse::class.java,
                )

            TmdbExceptions.from(
                httpStatus,
                errorResponse.statusCode,
                errorResponse.statusMessage,
                body,
            )
        } catch (exception: Exception) {
            TmdbExceptions.fromHttpStatus(
                httpStatus,
                "TMDB request failed with status code $httpStatus",
                body,
            )
        }

    private fun <T> mapBody(
        body: String,
        responseType: Class<T>,
        method: String,
        path: String,
    ): T =
        try {
            objectMapper.readValue(body, responseType)
        } catch (exception: Exception) {
            metricsRecorder.recordMappingFailed(
                method,
                path,
                responseType,
                exception,
            )

            throw TmdbMappingException(
                "Failed to map TMDB response to ${responseType.simpleName}",
                exception,
            )
        }

    private fun logNonSuccessfulResponse(
        httpStatus: Int,
        path: String,
        durationMillis: Long,
    ) {
        when {
            httpStatus == 429 -> {
                log.warn(
                    "TMDB rate limit response received: status={}, path={}, duration={}ms",
                    httpStatus,
                    path,
                    durationMillis,
                )
            }

            httpStatus >= 500 -> {
                log.warn(
                    "TMDB server error response received: status={}, path={}, duration={}ms",
                    httpStatus,
                    path,
                    durationMillis,
                )
            }
        }
    }

    private fun elapsedMillis(startedAt: Long): Long = (System.nanoTime() - startedAt) / 1_000_000

    private fun responseBytes(body: String?): Long = body?.toByteArray(StandardCharsets.UTF_8)?.size?.toLong() ?: 0L

    companion object {
        private val log = LoggerFactory.getLogger(JavaNetTmdbHttpClient::class.java)
    }
}
